### PR TITLE
protonmail-bridge-gui: patch source to support Qt 6.8

### LINKED ISSUE #350165: ProtonMail Bridge 3.14 fails to start on Nixos unstable with Qt 6.8
### Describe the bug
Apparently protonmail-bridge-gui has issues on systems with Qt 6.8 (i.e. recent unstable builds). The Proton team already knows it (https://github.com/ProtonMail/proton-bridge/issues/500) but they do seem to put the Qt update high on the priority list, citing distribution support. Opening this issue for all those wondering if it's their problem or if it's a wider / not-strictly-Nixos-related issue.

Also: would it be possible to "wrap the package" in a way that its dependencies (Qt 6.7) are met nonetheless? I do not know the inner workings of Nixos enough to have an answer to this myself.

### Steps To Reproduce
1. grab a Nixos Unstable and update it to current
2. install protonmail-bridge-gui
3. open it

### Expected behavior
ProtonMail bridge GUI working as usual (starting up, ending up in tray, serving mail to local mail clients).

### Screenshots
Screenshot from 2024-10-21 09-14-03

### Additional context
Trying to setup a workaround based on the CLI package, will add info to this issue on the result.

### Notify maintainers
@daniel-fahey 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.4, NixOS, 24.11 (Vicuna), 24.11pre694395.4c2fcb090b1f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.9`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by daniel-fahey**
Fixes https://github.com/NixOS/nixpkgs/issues/350165 (runtime errors encountered when using a version of the GUI built with Qt 6.8). Patches upstream's Qt Modeling Language (QML) files to support Qt 6.8. [Proton AG seem unlikely to update upstream to support Qt 6.8 anytime soon.](https://github.com/ProtonMail/proton-bridge/issues/500)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/pr/protonmail-bridge-gui/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/pr/protonmail-bridge-gui/package.nix:74]: Wouldn't it be easier to maintain a patch?

Comment (daniel-fahey) [reply pkgs/by-name/pr/protonmail-bridge-gui/package.nix:74]: I thought it would be "easier" (less work for me after each upstream update) to maintain a `sed` script than a big patch affecting dozens of files (that I would generate after running these commands on the new upstream source anyway).

I want to follow best practices though and would be happy to defer to your experience in the matter! I came across this [thread in the forum](https://discourse.nixos.org/t/when-to-use-substituteinplace-functions-vs-a-patch/11073/12) and this [blog post](https://drakerossman.com/blog/how-to-patch-a-package-source-on-nixos) that suggest it's better to use [`substituteInPlace`](https://nixos.org/manual/nixpkgs/unstable/#fun-substituteInPlace) instead of `find...-exec...sed -i...`, and Mic92's observation that
> A patch file usually has the advantage that it fails loudly when it no longer applies.

Let me have a think about the best way forward. If you have any other links re: best practices for this I'll read and consider them, as what I decide to do should be consistently applied across the derivation's entire `postPatch` section.

**Comment by daniel-fahey** [comment]
Looks like the next release will be patched to support building with Qt 6.8 https://github.com/ProtonMail/proton-bridge/pull/502#issuecomment-2442442521

**Review by Naxdy — COMMENTED** [review]
Given that https://github.com/ProtonMail/proton-bridge/pull/502 seems to have been accepted upstream, I think it makes sense to include the contents of that PR as a patch rather than using the sed approach.

File: pkgs/by-name/pr/protonmail-bridge-gui/package.nix
Comment (SuperSandro2000) [reply pkgs/by-name/pr/protonmail-bridge-gui/package.nix:74]: The sed is so complex that when it fails I couldn't fix it. A patch could be applied on the source tree and more easily be fixed there.

Comment (daniel-fahey) [reply pkgs/by-name/pr/protonmail-bridge-gui/package.nix]: I'll remember to KISS!

**Comment by daniel-fahey** [comment]
I've replaced my `sed` scripts with the Arch patches. For (my own) reference I made the patches using:
```
git remote add proton https://github.com/ProtonMail/proton-bridge.git
git fetch proton pull/502/head:pr-502
git format-patch master..pr-502 \
  --full-index \
  --stat \
  --summary \
  --cover-from-description=message \
  --base=auto \
  --relative=internal/frontend/bridge-gui/
```

**Review by Naxdy — APPROVED** [review]
Package builds & GUI works again :+1:

**Comment by aacebedo** [comment]
can we merge this if it works ?

**Comment by daniel-fahey** [comment]
Superseded by #359899

**Comment by Phr0stByte** [comment]
All of this is very confusing - is there a nixos 24.11 fix? If so, can we just get the fixed one in the 24.11 channels so we can simply just run an update?

**Comment by Phr0stByte** [comment]
> > All of this is very confusing - is there a nixos 24.11 fix? If so, can we just get the fixed one in the 24.11 channels so we can simply just run an update?
> 
> The backport was merged into the `release-24.11` branch yesterday #360778
> 
> It's not yet reached the `nixos-24.11` branch as of writing (according to https://nixpk.gs/pr-tracker.html?pr=360778)

Thanks. Using the flatpak which is working for now, until the fix makes the main nixos channels

**Comment by daniel-fahey** [comment]
It's just reached `nixos-24.11`

**Comment by Phr0stByte** [comment]
> It's just reached `nixos-24.11`

Confirmed - and working perfectly! Thanks!!!

**Pull request closed without merging**
